### PR TITLE
Support for sum with axis in nopython mode.

### PR DESCRIPTION
**Pull request by DrTodd13**

### CODE REVIEW

File: numba/targets/base.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
                        RegistryLoader)
 from numba import datamodel
 
+import traceback
```
Comment (sklam) [inline numba/targets/base.py]: left over from debugging?

File: numba/types/functions.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 from .abstract import *
 from .common import *
 
+import pdb
```
Comment (sklam) [inline numba/types/functions.py]: left over from debugging?

File: numba/typing/npydecl.py [hunk 3]
```diff
@@ -350,7 +350,8 @@ class Numpy_method_redirection(AbstractTemplate):
 
     def generic(self, args, kws):
         assert not kws
-        [arr] = args
+        #[arr] = args
```
Comment (stuartarchibald) [inline numba/typing/npydecl.py]: delete?

File: numba/errors.py [hunk 4]
```diff
@@ -221,6 +221,12 @@ def __init__(self, exception):
         self.old_exception = exception
 
 
+class RequireConstValue(TypingError):
+    """For signaling a function typing require constant value for some of
+    its argument
```
Comment (DrTodd13) [inline numba/errors.py]: Done.

File: numba/targets/arraymath.py [hunk 5]
```diff
@@ -21,6 +21,130 @@
...
+
+        def array_indexer(a, i):
+            return a[i]
+
+        # loop to fill the tuple
+        for i in range(nd):
+            dataidx = cgctx.get_constant(types.intp, i)
+            # compile and call array_indexer
+            data = cgctx.compile_internal(builder, array_indexer,
+                                          #shape_list.dtype(shape_list, types.intp),
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Removed.

File: numba/targets/arraymath.py [hunk 6]
```diff
@@ -21,6 +21,130 @@
...
+    # The function signature for this intrinsic.
+    funtion_sig = tupty(shape_list, shape_tuple)
+
+    def codegen(cgctx, builder, signature, args):
+        lltupty = cgctx.get_value_type(tupty)
+        # Create an empty int tuple.
+        tup = cgutils.get_null_value(lltupty)
+
+        # Get the shape list from the args and we don't need shape tuple.
+        [in_shape, _] = args
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: I think this is how Siu did it in his example.  I believe you are right though that it doesn't matter so it is a matter of style?  Personally, I like this style but let me know if you really want it changed and I'll change it.

File: numba/targets/arraymath.py [hunk 7]
```diff
@@ -21,6 +21,130 @@
...
+    before = axis_value
+    after  = nd - before - 1
+    types_list = ([types.slice2_type] * before) +  \
+                  [types.intp] +                   \
+                 ([types.slice2_type] * after)
+    
+    # Creates the output type of the function.
+    tupty = types.Tuple(types_list)
+    # Defines the signature of the intrinsic.
+    funtion_sig = tupty(shape_tuple, value, axis)
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Was a type originating in Siu's example.  I've fixed it.

File: numba/targets/arraymath.py [hunk 8]
```diff
@@ -21,6 +21,130 @@
...
+
+    def codegen(cgctx, builder, signature, args):
+        lltupty = cgctx.get_value_type(tupty)
+        # Create an empty indexing tuple.
+        tup = cgutils.get_null_value(lltupty)
+
+        # We only need value of the axis dimension here.
+        # The rest are constants defined above.
+        [_, value_arg, _] = args
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: See reply above.  I kind of like this style.

File: numba/targets/arraymath.py [hunk 9]
```diff
@@ -21,6 +21,130 @@
...
+
+        def create_full_slice():
+            return slice(None,None)
+
+        # loop to fill the tuple with slice(None,None) before
+        # the axis dimension.
+        for i in range(0, axis_value):
+            dataidx = cgctx.get_constant(types.intp, i)
+            # compile and call array_indexer
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Good catch.  Fixed.

File: numba/typing/arraydecl.py [hunk 10]
```diff
@@ -11,6 +11,8 @@
 from numba.typing import collections
 from numba.errors import TypingError
 
+import pdb
```
Comment (DrTodd13) [inline numba/typing/arraydecl.py]: Fixed.  Thought I did this already.  Second try then.

File: numba/typing/arraydecl.py [hunk 11]
```diff
@@ -574,8 +586,11 @@ def array_attribute_attachment(self, ary):
     install_array_method(fname, generic_homog)
 
 # Functions that return a machine-width type, to avoid overflows
-for fname in ["sum", "prod"]:
-    install_array_method(fname, generic_expand)
+#for fname in ["sum", "prod"]:
+#    install_array_method(fname, generic_expand)
```
Comment (DrTodd13) [inline numba/typing/arraydecl.py]: Deleted.

File: numba/typing/npydecl.py (hunk 3 above)
Comment (DrTodd13) [inline numba/typing/npydecl.py]: Deleted.

File: numba/targets/arraymath.py [hunk 12]
```diff
@@ -21,6 +21,130 @@
...
+        # We only need value of the axis dimension here.
+        # The rest are constants defined above.
+        [_, value_arg, _] = args
+
+        def create_full_slice():
+            return slice(None,None)
+
+        # loop to fill the tuple with slice(None,None) before
+        # the axis dimension.
+        for i in range(0, axis_value):
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Did it this way to be consistent with the range call slightly below that.  Can change if you really want.

File: numba/targets/arraymath.py [hunk 13]
```diff
@@ -40,6 +164,67 @@ def array_sum_impl(arr):
...
+def array_sum_axis(context, builder, sig, args):
+    """
+    The third parameter to gen_index_tuple that generates the indexing
+    tuples has to be a const so we can't just pass "axis" through since
+    that isn't const.  We can check for specific values and have 
+    different instances that do take consts.  Supporting only four
+    dimensional arrays for now.
+    """
+
+    # sum_extend defines the return type for sum with axis.
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Was a typo.  Should be sum_expand.  Added a reference to where the return type is computed for sum with axis is computed.

File: numba/targets/arraymath.py [hunk 14]
```diff
@@ -40,6 +164,67 @@ def array_sum_impl(arr):
...
     return impl_ret_borrowed(context, builder, sig.return_type, res)
 
+@lower_builtin(np.sum, types.Array, types.int64)
+@lower_builtin("array.sum", types.Array, types.int64)
+def array_sum_axis(context, builder, sig, args):
+    """
+    The third parameter to gen_index_tuple that generates the indexing
+    tuples has to be a const so we can't just pass "axis" through since
+    that isn't const.  We can check for specific values and have 
+    different instances that do take consts.  Supporting only four
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Fixed.

File: numba/targets/arraymath.py [hunk 15]
```diff
@@ -40,6 +164,67 @@ def array_sum_impl(arr):
...
+        # Convert the shape of the input array to a list.
+        ashape = list(arr.shape)
+        # Get the length of the axis dimension.
+        axis_len = ashape[axis]
+        # Remove the axis dimension from the list of dimensional lengths.
+        ashape.pop(axis)
+        # Convert this shape list back to a tuple using above intrinsic.
+        ashape_without_axis = _create_tuple_result_shape(ashape, arr.shape)
+        # Tuple needed here to create output array with correct size.
+        result = np.full(ashape_without_axis, zero, type(zero))
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Done.

File: numba/targets/arraymath.py [hunk 16]
```diff
@@ -40,6 +164,67 @@ def array_sum_impl(arr):
...
+                index_tuple2 = _gen_index_tuple(arr.shape, axis_index, 1)
+                result += arr[index_tuple2]
+            elif axis == 2:
+                index_tuple3 = _gen_index_tuple(arr.shape, axis_index, 2)
+                result += arr[index_tuple3]
+            elif axis == 3:
+                index_tuple4 = _gen_index_tuple(arr.shape, axis_index, 3)
+                result += arr[index_tuple4]
+            else:
+                raise ValueError("Numba does not support sum with axis \
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Done.

File: numba/targets/arraymath.py [hunk 17]
```diff
@@ -21,6 +21,130 @@
 from numba.typing import signature
 from .arrayobj import make_array, load_item, store_item, _empty_nd_impl
 
+import traceback
```
Comment (DrTodd13) [inline numba/targets/arraymath.py]: Done.

**Comment by seibert** [comment]
Did you push your changes to the branch?  I'm not seeing the updates mentioned above.

**Comment by DrTodd13** [comment]
Just pushed what I have.  The tests I added for negative axis and axis > 3 are somehow causing the GC allocs==free assertion to fail.

**Comment by DrTodd13** [comment]
Can one of the Continuum folks look at these failures?  For each test of an invalid axis argument, there seems to be one object getting allocated that isn't being freed.  I pushed the axis check to basically the first thing in the function so I don't know where an object is being created.

File: numba/targets/arraymath.py (hunk 6 above)
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: I'm happy either way :) I just spot it out of habit, no change needed.

File: numba/targets/arraymath.py (hunk 12 above)
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: It's fine to leave.

**Comment by stuartarchibald** [comment]
Thanks for the fixups, much appreciated. I'll take a look at what's going on with the ref leaks.

RE: https://github.com/numba/numba/pull/2482/commits/00f60bab0e9852e9107be9077019c51c22456f8c `np.take` was added in https://github.com/numba/numba/pull/2490 (very recent).

**Comment by stuartarchibald** [comment]
I think because the exceptions are coming from compiled code it's likely it's a valid leak. https://github.com/stuartarchibald/numba/commit/f012f072b9fd59ae760eaa9b35f2d2d9d3467ed3 sorts it by splitting the leaking exception raising calls into a test with the reference count assertion disabled, you are welcome to use or rework as you see fit.

**Comment by DrTodd13** [comment]
For numpysupported.rst, do you want sum in both lists or just the second one?

File: numba/targets/arraymath.py [hunk 18]
```diff
@@ -40,6 +158,69 @@ def array_sum_impl(arr):
...
+    that isn't const.  We can check for specific values and have 
+    different instances that do take consts.  Supporting axis summation
+    only up to the fourth dimension for now.
+    """
+
+    # typing/arraydecl.py:sum_expand defines the return type for sum with axis. 
+    # It is one dimension less than the input array.
+    zero = sig.return_type.dtype(0)
+    # Again, the result array is one less than the input array dimensions.
+    ndim = sig.args[0].ndim - 1
```
Comment (sklam) [inline numba/targets/arraymath.py]: `ndim` is not used

**Comment by sklam** [comment]
This PR looks good.  I opened https://github.com/numba/numba/pull/2503 (still WIP) to remove the hacky code I added.

**Comment by seibert** [comment]
Thanks for the PR!  I'm going to merge, and Siu can refactor the const code in #2503.

**Pull request merged**